src/treq/test/test_treq_integration.py

```diff
@@ -241,6 +243,156 @@ def test_failed_basic_auth(self):
+        """
+            Test successful Digest authentication with multiple qop types
+        :return:
+        """
+        response = yield self.get('/digest-auth/undefined/treq/treq',
+                                  auth=HTTPDigestAuth('treq', 'treq'))
+        self.assertEqual(response.code, 200)
+        yield print_response(response)
+        json = yield treq.json_content(response)
+        self.assertTrue(json['authenticated'])
+        self.assertEqual(json['user'], 'treq')
+
+    @inlineCallbacks
+    def test_digest_auth_multiple_calls(self):
+        """
+            Test proper Digest authentication credentials caching
+        """
+
+        # A mutable holder for call counter
+        agent_request_call_storage = {
+            'c': 0,
+            'i': []
+        }
+
+        # Original Agent request call
+        agent_request_orig = Agent.request
+
+        def agent_request_patched(*args, **kwargs):
+            """
+                Patched Agent.request function,
+                that inscreaces call count on every HTTP request
+                and appends
+            """
+            response_deferred = agent_request_orig(*args, **kwargs)
+            agent_request_call_storage['c'] += 1
+            agent_request_call_storage['i'].append((args, kwargs))
+            return response_deferred
+
+        patcher = MonkeyPatcher(
+            (Agent, 'request', agent_request_patched)
+        )
+        patcher.patch()
+
+        auth = HTTPDigestAuth('treq-digest-auth-multiple', 'treq')
+
+        response1 = yield self.get(
+            '/digest-auth/auth/treq-digest-auth-multiple/treq',
+            auth=auth
+        )
+        self.assertEqual(response1.code, 200)
+        yield print_response(response1)
+        json1 = yield treq.json_content(response1)
+
+        # Assume we did two actual HTTP requests - one to obtain credentials
+        # and second is original request with authentication
+        self.assertEqual(
+            agent_request_call_storage['c'],
+            2
+        )
+        headers_for_second_request = agent_request_call_storage['i'][1][0][3]
+        self.assertIn(
+            b'Authorization',
+            dict(headers_for_second_request.getAllRawHeaders())
+        )
+
+        response2 = yield self.get(
+            '/digest-auth/auth/treq-digest-auth-multiple/treq',
+            auth=auth,
+            cookies=response1.cookies()
+        )
+        self.assertEqual(response2.code, 200)
+        yield print_response(response2)
+        json2 = yield treq.json_content(response2)
+        self.assertTrue(json1['authenticated'])
+        self.assertEqual(json1['user'], 'treq-digest-auth-multiple')
+
+        # Assume that responses are the same
+        self.assertEqual(json1, json2)
+
+        # Assume we need only one call to obtain second response
+        self.assertEqual(
+            agent_request_call_storage['c'],
+            3
+        )
+        patcher.restore()
```

Comment:
This needs to be an `addCleanup` added immediately after starting so the patch can't leak on test failure (or `self.patch` does that automatically).

Reply:
switched to `self.patch`